azurerm/internal/services/dns/dns_zone_data_source.go

```diff
@@ -109,26 +107,33 @@ func dataSourceArmDnsZoneRead(d *schema.ResourceData, meta interface{}) error {
 	return tags.FlattenAndSet(d, resp.Tags)
 }
 
-func findZone(client *dns.ZonesClient, rgClient *resources.GroupsClient, ctx context.Context, name string) (dns.Zone, string, error) {
-	groups, err := rgClient.List(ctx, "", nil)
+func findZone(client *dns.ZonesClient, ctx context.Context, name string) (dns.Zone, string, error) {
+	zonesIterator, err := client.ListComplete(ctx, nil)
 	if err != nil {
-		return dns.Zone{}, "", fmt.Errorf("Error listing Resource Groups: %+v", err)
+		return dns.Zone{}, "", fmt.Errorf("listing DNS Zones: %+v", err)
 	}
 
-	for _, g := range groups.Values() {
-		resourceGroup := *g.Name
-
-		zones, err := client.ListByResourceGroup(ctx, resourceGroup, nil)
-		if err != nil {
-			return dns.Zone{}, "", fmt.Errorf("Error listing DNS Zones (Resource Group: %s): %+v", resourceGroup, err)
-		}
-
-		for _, z := range zones.Values() {
-			if *z.Name == name {
-				return z, resourceGroup, nil
+	var found *dns.Zone
+	for zonesIterator.NotDone() {
+		zone := zonesIterator.Value()
+		if zone.Name != nil && *zone.Name == name {
+			if found != nil {
+				return dns.Zone{}, "", fmt.Errorf("found multiple DNS zones with name %q, please specify the resource group", name)
 			}
+			found = &zone
 		}
+		if err := zonesIterator.NextWithContext(ctx); err != nil {
+			return dns.Zone{}, "", fmt.Errorf("listing DNS Zones: %+v", err)
+		}
+	}
+
+	if found == nil || found.ID == nil {
+		return dns.Zone{}, "", fmt.Errorf("could not find DNS zone with name: %q", name)
 	}
 
-	return dns.Zone{}, "", nil
+	id, err := parse.DnsZoneID(*found.ID)
+	if err != nil {
+		return dns.Zone{}, "", fmt.Errorf("DNS zone id not valid: %+v", err)
+	}
+	return *found, id.ResourceGroup, nil
```

Comment:
this should be de-referenced anyway - so it should be fine:

```suggestion
	return found, id.ResourceGroup, nil
```